File: Task/Largest-prime-factor/Haskell/largest-prime-factor.hs
isPrime :: Integer -> Bool
isPrime n
  | n < 2 = False
  | n == 2 = True
  | otherwise = 0 == length [i | i <- [2..floor(sqrt (fromIntegral n))], n `mod` i == 0]

largestPrimeFactor :: Integer -> Integer
largestPrimeFactor n
  | n < 2 = 1
  | otherwise = max smallestPrime (largestPrimeFactor(n `div` smallestPrime))
  where
    setOfAllPrimes = [i | i <- [2..], isPrime i]
    isNotDivisible x y = 0 /= x `mod` y
    smallestPrime = head (dropWhile (isNotDivisible n) setOfAllPrimes)

x = 600851475143
main = print(largestPrimeFactor x)

File: Task/Largest-prime-factor/C/largest-prime-factor.c
#include <stdio.h>
#include <stdlib.h>

int isprime( long int n ) {
    int i=3;
    if(!(n%2)) return 0;
    while( i*i < n ) {
        if(!(n%i)) return 0;
        i+=2;
    }
    return 1;
}

int main(void) {
    long int n=600851475143, j=3;

    while(!isprime(n)) {
        if(!(n%j)) n/=j;
        j+=2;
    }
    printf( "%ld\n", n );
    return 0;
}
